rfd/headless-login.md

```diff
@@ -0,0 +1,367 @@
+
+4. **The Server must verify that the client requesting certificates is the only client that can use the certificates.**
+
+Depending on the headless authentication API flow, it may be split into multiple requests. It is crucial that the server verifies that the client who requested headless authentication is the only client that can use the resulting credentials. This will be accomplished with a PKCE-like flow, where the client provides their public key in the initial request, and the server ultimately issues certificates for the original client public key. Additionally, the user will be asked to sign for the public key before the server issues the certificates (TODO: is this possible without x509 certificate + TLS handshake)
+
+Note: Ideally, we would ensure that the only client that can *receive* certificates is the requesting client, but ensuring the PKCE flow above is enough to defend against attacks since the certificates are useless without the client's private key. If we want to remove the possibility of an attack attempting to intercept the certificates, we can use an HTTP websocket or gRPC stream endpoint to encapsulate the headless authentication request into a single API request. This would require some additional implementation details to be flushed out, so it is not in this initial draft design.
+
+### Headless authentication overview
+
+The headless authentication flow is shown below:
+
+```mermaid
+sequenceDiagram
+    participant Local Machine
+    participant Remote Machine
+    participant Teleport Proxy
+    participant Teleport Auth
+
+    Note over Remote Machine: tsh --headless ssh user@node01
+    Remote Machine->>+Teleport Proxy: initiate headless login
+    Teleport Proxy->>Teleport Auth: save request with new uuid
+    Note over Teleport Auth: Request: {id, user, publicKey, ...}
+    Teleport Auth->>Teleport Proxy: request ID
+    Teleport Proxy->>-Remote Machine: request ID
+    Note over Remote Machine: print command + URL
+
+    Remote Machine->>+Teleport Proxy: request signed user certs<br/>for request ID
+    Note over Local Machine: user copies command / URL<br/>to local terminal / browser
+    Local Machine->>+Teleport Proxy: initiate headless request approval
+    Teleport Proxy->>Local Machine: MFA challenge
+    Note over Local Machine: user taps YubiKey
+    Local Machine->>Teleport Proxy: signed MFA challenge response
+    Teleport Proxy-->>-Local Machine: success/failure
+
+    Teleport Proxy->>Teleport Auth: Request signed user certificates
+    Teleport Auth->>Teleport Proxy: Issue signed user certificates<br/>for the client's public key 
+    Teleport Proxy->>-Remote Machine: User certificates
+    Note over Remote Machine: Connect to user@node01
```

Comment:
Should we somehow verify if the authentication data matches the data from the headless request?

Reply:
Do you mean [this](https://github.com/gravitational/teleport/blob/b26a3c39473026572370042b3edb7076a4a25c3f/rfd/headless-login.md#post-webapiloginheadlessfinish):

> However, the request will be denied the the request ID, user, and public key do not match the HeadlessLoginRequest saved on the auth server.

or something else?

Reply:
I think this is enough. I'd be just more explicit about the public key part and say that the request has to be signed by a matching private key.